### PR TITLE
ICEBERG-2559: Add utility methods

### PR DESCRIPTION
**Pull request by karuppayya**
This is a precursor for PR https://github.com/apache/iceberg/pull/2415
Check comment https://github.com/apache/iceberg/pull/2415#discussion_r627834645 for more details

### CODE REVIEW

**Comment by karuppayya** [comment]
cc: @aokolnychyi @RussellSpitzer @flyrain @kbendick for review. Thanks

**Review by kbendick — APPROVED** [review]
This looks good to me.

Small possible point of confusion on the doc comment for `miscMetadataFiles` but the logic all looks sound to me. Thanks @karuppayya.

File: core/src/main/java/org/apache/iceberg/MetadataLocationUtils.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+        metadataLogEntries.stream().map(TableMetadata.MetadataLogEntry::file)
+            .collect(Collectors.toList());
+    if (previousMetadataFiles.size() > 0) {
+      metaFiles.addAll(previousMetadataFiles);
+      if (isRecursive) {
+        for (String fileName : previousMetadataFiles) {
+          // Find the first existent metadata json file and recurse
+          if (io.newInputFile(fileName).exists()) {
+            miscMetadataFiles(fileName, metaFiles, io, isRecursive);
+            break;
```
Comment (flyrain) [inline core/src/main/java/org/apache/iceberg/MetadataLocationUtils.java:68]: Why `break`? When I look at the test `testMiscMetadataFilesWithMissingFiles`, what if v3.metadata.json be the first one in the loop?  In that case, v1 isn't reachable since v3 has been deleted, and the loop will be terminated because of the `break`.

File: core/src/main/java/org/apache/iceberg/MetadataLocationUtils.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Returns all the path locations of all Manifest Lists for a given list of snapshots
```
Comment (flyrain) [inline core/src/main/java/org/apache/iceberg/MetadataLocationUtils.java:76]: I'd expect method parameter to be a list of snapshots. Is this method for a given table?

**Comment by aokolnychyi** [comment]
Sorry for the delay @karuppayya, let me take a look now.

File: core/src/main/java/org/apache/iceberg/MetadataLocationUtils.java [hunk 3]
```diff
@@ -58,17 +53,22 @@ private static void miscMetadataFiles(String metadataFileLocation, Set<String> m
...
+    TableMetadata metadata = TableMetadataParser.read(io, metadataFileLocation);
+    List<TableMetadata.MetadataLogEntry> metadataLogEntries = metadata.previousFiles();
+    List<String> previousMetadataFiles =
+        metadataLogEntries.stream().map(TableMetadata.MetadataLogEntry::file)
+            .collect(Collectors.toList());
+    if (previousMetadataFiles.size() > 0) {
       metaFiles.addAll(previousMetadataFiles);
-      if (isRecursive && previousMetadataFiles.size() > 0) {
-        miscMetadataFiles(previousMetadataFiles.get(0), metaFiles, io, isRecursive);
+      if (isRecursive) {
```
[the remark this answers is not in the document]
Comment (karuppayya) [reply core/src/main/java/org/apache/iceberg/MetadataLocationUtils.java:63]: In the above case, we will first pick the latest version(ie file 3), add all the files in file 3.
Pick the oldest exist file from the list of files in file 3, and recurse with that file.
```
      if (isRecursive) {
        for (String fileName : previousMetadataFiles) {
          // Find the first existent metadata json file and recurse
          if (io.newInputFile(fileName).exists()) {
            miscMetadataFiles(fileName, metaFiles, io, isRecursive);
            break;
          }
        }
      }
```
We dont queue up all the files present in a particular file. The max number of files queued up will be 1

File: core/src/main/java/org/apache/iceberg/MetadataLocationUtils.java (hunk 2 above)
Comment (karuppayya) [reply core/src/main/java/org/apache/iceberg/MetadataLocationUtils.java:76]: sorry, missed to fix the javadoc. Thanks!!

File: core/src/main/java/org/apache/iceberg/MetadataLocationUtils.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (RussellSpitzer) [reply core/src/main/java/org/apache/iceberg/MetadataLocationUtils.java:63]: Maybe we should have that be a "findFirst" then instead of the break? To make things a bit more straightforward?

[the remark this answers is not in the document]
Comment (flyrain) [reply core/src/main/java/org/apache/iceberg/MetadataLocationUtils.java:63]: Make sense to get the first item more explicitly. Another concern, is there a chance that order is broken?

[the remark this answers is not in the document]
Comment (flyrain) [reply core/src/main/java/org/apache/iceberg/MetadataLocationUtils.java:63]: oh, wait, to get the first isn't correct here. Image you have v1 v2 v3 and v4, METADATA_PREVIOUS_VERSIONS_MAX is 2. You can delete v2 in v4. v2 is the first one, if you only search for the first one, you are going to miss things in v3.metadata.json, which points to v1 and v2.

[the remark this answers is not in the document]
Comment (karuppayya) [reply core/src/main/java/org/apache/iceberg/MetadataLocationUtils.java:63]: We need to find the first existent file, like in my previous logic. 
But there will be issues with that as well. There can be unreachable files like in this test case. https://github.com/apache/iceberg/pull/2564/files#diff-efb7c769faae3f19c11382d2dc4c53a69074265a87f7291f8430ad91410d48e2R141

[the remark this answers is not in the document]
Comment (flyrain) [reply core/src/main/java/org/apache/iceberg/MetadataLocationUtils.java:63]: What is the use case of getting all metadata.json files? How about listing all *metadata.json files in the dir?

[the remark this answers is not in the document]
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/MetadataLocationUtils.java:63]: I think we can try to load previous metadata files but give up if we cannot do that. It should be okay to leave those files too. In case if Iceberg owns the location, we will drop the entire location after that too. If not, we can leave them be.

Listing may not be an option as the same location may contain files for multiple tables. In HiveCatalog, we have table UUID as part of the file name while we don't have that in HadoopCatalog for example. 

If it is too complicated, we can drop the recursion too and I say we only clean up files currently referenced.

File: core/src/main/java/org/apache/iceberg/MetadataLocationUtils.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+  }
+
+  /**
+   * Returns all the path locations of all Manifest Lists for a given table
+   * @param table table for which manifestList needs to be fetched
+   * @return the paths of the Manifest Lists
+   */
+  public static List<String> manifestListPaths(Table table) {
+    Iterable<Snapshot> snapshots = table.snapshots();
+    List<String> manifestLists = new ArrayList<>();
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/MetadataLocationUtils.java:82]: nit: I have nothing against constructing `ArrayList`s directly but historically Iceberg uses `Lists.newArrayList()` so I think it makes sense to be consistent with the rest of the codebase.

Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/MetadataLocationUtils.java:82]: nit: `manifestLists` -> `manifestListLocations`

File: core/src/main/java/org/apache/iceberg/MetadataLocationUtils.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+package org.apache.iceberg;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.iceberg.io.FileIO;
+
+public class MetadataLocationUtils {
```
[the remark this answers is not in the document]
Comment (flyrain) [reply core/src/main/java/org/apache/iceberg/MetadataLocationUtils.java:29]: `MetadataFileUtil`?

File: core/src/main/java/org/apache/iceberg/ReachableFileUtils.java [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    TableMetadata tableMetadata = ops.current();
+    metadataFileLocations.add(tableMetadata.metadataFileLocation());
+    metadataFileLocations(tableMetadata, metadataFileLocations, ops.io(), recursive);
+    return metadataFileLocations;
+  }
+
+  private static void metadataFileLocations(TableMetadata metadata, Set<String> metaFiles,
+                                            FileIO io, boolean isRecursive) {
+    List<TableMetadata.MetadataLogEntry> metadataLogEntries = metadata.previousFiles();
+    List<String> previousMetadataFiles =
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/ReachableFileUtils.java:68]: This creates a temp list that gets added to the set immediately. Let's not add the overhead.

File: core/src/main/java/org/apache/iceberg/ReachableFileUtils.java [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+        metadataLogEntries.stream().map(TableMetadata.MetadataLogEntry::file)
+            .collect(Collectors.toList());
+    if (previousMetadataFiles.size() > 0) {
+      metaFiles.addAll(previousMetadataFiles);
+      // Find the first existent metadata json file and recurse
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/ReachableFileUtils.java:73]: I think we better give up if the oldest one is not reachable. Trying metadata files one by one will send a request each time. If the lineage is broken, that's not our fault and it will not impact correctness.

File: core/src/main/java/org/apache/iceberg/ReachableFileUtils.java [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+  public static Set<String> metadataFileLocations(Table table, boolean recursive) {
+    Set<String> metadataFileLocations = new HashSet<>();
+    TableOperations ops = ((HasTableOperations) table).operations();
+    TableMetadata tableMetadata = ops.current();
+    metadataFileLocations.add(tableMetadata.metadataFileLocation());
+    metadataFileLocations(tableMetadata, metadataFileLocations, ops.io(), recursive);
+    return metadataFileLocations;
+  }
+
+  private static void metadataFileLocations(TableMetadata metadata, Set<String> metaFiles,
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/ReachableFileUtils.java:65]: I think I still prefer something like [this](https://github.com/apache/iceberg/pull/2564#discussion_r629842136) without a separate method.

File: core/src/main/java/org/apache/iceberg/ReachableFileUtils.java [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+          } catch (Exception e) {
+            LOG.error("Failed to load {}", metadataFileLocation, e);
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Returns all the path locations of all Manifest Lists for a given table
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/ReachableFileUtils.java:89]: nit: `Returns locations of manifest lists in a table.`

Let's add an empty line before params too.

File: spark/src/main/java/org/apache/iceberg/spark/actions/BaseSparkAction.java [hunk 10]
```diff
@@ -109,40 +110,6 @@ protected JobGroupInfo newJobGroupInfo(String groupId, String desc) {
     return new JobGroupInfo(groupId + "-" + JOB_COUNTER.incrementAndGet(), desc, false);
   }
 
-  /**
-   * Returns all the path locations of all Manifest Lists for a given list of snapshots
-   * @param snapshots snapshots
-   * @return the paths of the Manifest Lists
-   */
-  private List<String> getManifestListPaths(Iterable<Snapshot> snapshots) {
```
Comment (flyrain) [inline spark/src/main/java/org/apache/iceberg/spark/actions/BaseSparkAction.java:117]: Can we remove the same method in class `BaseAction`? This applies to `getOtherMetadataFilePaths` as well.

File: core/src/main/java/org/apache/iceberg/ReachableFileUtil.java [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+  }
+
+  /**
+   * Returns the location of the version.text file
+   *
+   * @param ops tableOperation for which version.text path needs to be retrieved
+   * @return the location of the version hint file
+   */
+  public static String versionHintLocation(TableOperations ops) {
+    return ops.metadataFileLocation("version-hint.text");
```
Comment (flyrain) [inline core/src/main/java/org/apache/iceberg/ReachableFileUtil.java:44]: Per @kbendick suggested, it is a good time to make a constant for "version-hint.text". But I'm OK to do this in another PR.

File: core/src/main/java/org/apache/iceberg/MetadataTableUtils.java [hunk 12]
```diff
@@ -23,6 +23,7 @@
 import org.apache.iceberg.exceptions.NoSuchTableException;
 
 public class MetadataTableUtils {
```
Comment (flyrain) [inline core/src/main/java/org/apache/iceberg/MetadataTableUtils.java:25]: Do we need this class?

File: core/src/main/java/org/apache/iceberg/ReachableFileUtil.java [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+  public static Set<String> metadataFileLocations(TableOperations ops, boolean recursive) {
+    Set<String> metadataFileLocations = Sets.newHashSet();
+    TableMetadata tableMetadata = ops.current();
+    metadataFileLocations.add(tableMetadata.metadataFileLocation());
+    metadataFileLocations(tableMetadata, metadataFileLocations, ops.io(), recursive);
+    return metadataFileLocations;
+  }
+
+  private static void metadataFileLocations(TableMetadata metadata, Set<String> metaFiles,
+                                            FileIO io, boolean isRecursive) {
```
Comment (flyrain) [inline core/src/main/java/org/apache/iceberg/ReachableFileUtil.java:64]: A nit: let's be consistent of naming. isRecursive -> recursive.

File: core/src/main/java/org/apache/iceberg/ReachableFileUtil.java [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+    List<TableMetadata.MetadataLogEntry> metadataLogEntries = metadata.previousFiles();
+    if (metadataLogEntries.size() > 0) {
+      for (TableMetadata.MetadataLogEntry metadataLogEntry : metadataLogEntries) {
+        metaFiles.add(metadataLogEntry.file());
+      }
+      if (isRecursive) {
```
Comment (flyrain) [inline core/src/main/java/org/apache/iceberg/ReachableFileUtil.java:70]: Per off-line discussion, we are trying to find the first existing metadata.json file while recursive searching. Let's take the test as an example,  there are v1 v2 v3 and v4, METADATA_PREVIOUS_VERSIONS_MAX is 2. Inside the v4, you have v2 and v3, and v2 is the first item, if v2 is there, open v2 and continue searching, but if v2.metadata.json is deleted, we can open the second item(v3.metadata.json) and continue searching. It's simple logic-wise and perf is also good since we don't open the same file multiple times.

Comment (karuppayya) [reply core/src/main/java/org/apache/iceberg/ReachableFileUtil.java:70]: I was doing doing that initially. Based on https://github.com/apache/iceberg/pull/2564#discussion_r630695415, I am now checking the only the oldest file and bailing out if not existent.

Comment (flyrain) [reply core/src/main/java/org/apache/iceberg/ReachableFileUtil.java]: Thanks @karuppayya. I think your initial way makes sense, for implementation, I'd suggest to wrap the logic into a small method like `findFirstExistedJson`. Let's hear thoughts from @aokolnychyi.

File: core/src/main/java/org/apache/iceberg/ReachableFileUtils.java (hunk 8 above)
Comment (karuppayya) [reply core/src/main/java/org/apache/iceberg/ReachableFileUtils.java]: The original method takes `TableOperations` as input but the recursion happens on `TableMetadata`. Hence I think a second method is required. Please Letting me know if i am missing something here

File: core/src/main/java/org/apache/iceberg/ReachableFileUtil.java [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+  public static String versionHintLocation(TableOperations ops) {
+    return ops.metadataFileLocation(Util.VERSION_HINT_TXT_FILENAME);
+  }
+
+  /**
+   * Returns locations of JSON metadata files in a table.
+   *
+   * @param ops       TableOperations to get JSON metadata files from
+   * @param recursive When true, recursively retrieves all the reachable JSON metadata files.
+   *                 When false, gets the all the JSON metadata files only from the current metadata.
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/ReachableFileUtil.java:53]: nit: can we align both `when`?

File: core/src/main/java/org/apache/iceberg/ReachableFileUtil.java [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+    metadataFileLocations.add(tableMetadata.metadataFileLocation());
+    metadataFileLocations(tableMetadata, metadataFileLocations, ops.io(), recursive);
+    return metadataFileLocations;
+  }
+
+  private static void metadataFileLocations(TableMetadata metadata, Set<String> metaFiles,
+                                            FileIO io, boolean recursive) {
+    List<TableMetadata.MetadataLogEntry> metadataLogEntries = metadata.previousFiles();
+    if (metadataLogEntries.size() > 0) {
+      for (TableMetadata.MetadataLogEntry metadataLogEntry : metadataLogEntries) {
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/ReachableFileUtil.java:68]: Can we add an import for `MetadataLogEntry` to shorten the lines?

File: core/src/main/java/org/apache/iceberg/ReachableFileUtil.java [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+        metaFiles.add(metadataLogEntry.file());
+      }
+      if (recursive) {
+        String metadataFileLocation = metadataLogEntries.get(0).file();
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/ReachableFileUtil.java:72]: Would it make sense to name it like `oldestMetadataFileLocation` or `previousMetadataFileLocation`?

File: core/src/main/java/org/apache/iceberg/ReachableFileUtil.java [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+        try {
+          TableMetadata newMetadata = TableMetadataParser.read(io, metadataFileLocation);
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/ReachableFileUtil.java:74]: nit: `newMetadata` -> `previousMetadata` or something?

File: core/src/main/java/org/apache/iceberg/ReachableFileUtil.java [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+   */
+  public static Set<String> metadataFileLocations(TableOperations ops, boolean recursive) {
+    Set<String> metadataFileLocations = Sets.newHashSet();
+    TableMetadata tableMetadata = ops.current();
+    metadataFileLocations.add(tableMetadata.metadataFileLocation());
+    metadataFileLocations(tableMetadata, metadataFileLocations, ops.io(), recursive);
+    return metadataFileLocations;
+  }
+
+  private static void metadataFileLocations(TableMetadata metadata, Set<String> metaFiles,
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/ReachableFileUtil.java:64]: nit: `metaFiles` -> `metadataFileLocations`

File: core/src/test/java/org/apache/iceberg/util/TestReachableFileUtil.java [hunk 20]
```diff
@@ -0,0 +1,137 @@
...
+
+  private Table table;
+
+  @Before
+  public void setupTableLocation() throws Exception {
+    File tableDir = temp.newFolder();
+    String tableLocation = tableDir.toURI().toString();
+    this.table = TABLES.create(SCHEMA, SPEC, Maps.newHashMap(), tableLocation);
+  }
+
```
Comment (aokolnychyi) [inline core/src/test/java/org/apache/iceberg/util/TestReachableFileUtil.java:77]: nit: extra line

File: spark/src/main/java/org/apache/iceberg/spark/actions/BaseSparkAction.java (hunk 10 above)
Comment (aokolnychyi) [reply spark/src/main/java/org/apache/iceberg/spark/actions/BaseSparkAction.java:117]: Yeah, we should clean up `BaseAction`. I am fine doing that in a follow-up PR too. @karuppayya, could you submit one when the current is in?

**Review by aokolnychyi — COMMENTED** [review]
A few minor comments and should be good to go. Thanks, @karuppayya!

File: core/src/main/java/org/apache/iceberg/ReachableFileUtil.java (hunk 14 above)
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/ReachableFileUtil.java]: I am a bit worried about the number of requests we can send that way. I feel like if the lineage broken then it is fine to give up. That being said, I am also fine if both of you think it is going to be helpful. I agree a separate method should simplify it a bit.

Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/ReachableFileUtil.java]: Leave this up to you, @karuppayya @flyrain.

File: core/src/main/java/org/apache/iceberg/ReachableFileUtil.java (hunk 16 above)
Comment (karuppayya) [reply core/src/main/java/org/apache/iceberg/ReachableFileUtil.java]: `MetadataLogEntry` is  a static inner class, we would not be able to do a static import of the class.

File: core/src/main/java/org/apache/iceberg/ReachableFileUtil.java [hunk 21]
```diff
@@ -0,0 +1,100 @@
...
+  private ReachableFileUtil() {
+  }
+
+  /**
+   * Returns the location of the version.text file
+   *
+   * @param ops tableOperation for which version.text path needs to be retrieved
+   * @return the location of the version hint file
+   */
+  public static String versionHintLocation(TableOperations ops) {
```
[the remark this answers is not in the document]
Comment (karuppayya) [reply core/src/main/java/org/apache/iceberg/ReachableFileUtil.java]: I have refactored to use `Table` objects

File: core/src/main/java/org/apache/iceberg/ReachableFileUtil.java (hunk 14 above)
Comment (karuppayya) [reply core/src/main/java/org/apache/iceberg/ReachableFileUtil.java]: Add a additional method to find the first existent metadata and using it figure the metadata file locations.

**Review by flyrain — APPROVED** [review]
+1

File: core/src/main/java/org/apache/iceberg/ReachableFileUtil.java [hunk 22]
```diff
@@ -0,0 +1,114 @@
...
+    TableOperations ops = ((HasTableOperations) table).operations();
+    TableMetadata tableMetadata = ops.current();
+    metadataFileLocations.add(tableMetadata.metadataFileLocation());
+    metadataFileLocations(tableMetadata, metadataFileLocations, ops.io(), recursive);
+    return metadataFileLocations;
+  }
+
+  private static void metadataFileLocations(TableMetadata metadata, Set<String> metadataFileLocations,
+                                            FileIO io, boolean recursive) {
+    List<TableMetadata.MetadataLogEntry> metadataLogEntries = metadata.previousFiles();
```
Comment (aokolnychyi) [inline core/src/main/java/org/apache/iceberg/ReachableFileUtil.java:69]: nit: let's import `MetadataLogEntry` directly to shorten the lines.

File: spark/src/main/java/org/apache/iceberg/spark/actions/BaseSparkAction.java [hunk 23]
```diff
@@ -167,19 +133,21 @@ protected Table newStaticTable(TableMetadata metadata, FileIO io) {
...
-    List<String> manifestLists = getManifestListPaths(table.snapshots());
+    List<String> manifestLists = ReachableFileUtil.manifestListLocations(table);
     return spark.createDataset(manifestLists, Encoders.STRING()).toDF("file_path");
   }
 
-  protected Dataset<Row> buildOtherMetadataFileDF(TableOperations ops) {
-    List<String> otherMetadataFiles = getOtherMetadataFilePaths(ops);
-    return spark.createDataset(otherMetadataFiles, Encoders.STRING()).toDF("file_path");
+  protected Dataset<Row> buildOtherMetadataFileDF(Table table) {
+    Set<String> otherMetadataFiles = ReachableFileUtil.metadataFileLocations(table, false);
```
Comment (aokolnychyi) [inline spark/src/main/java/org/apache/iceberg/spark/actions/BaseSparkAction.java:141]: nit: I think it is a little confusing to call metadata file locations as otherMetadataFiles and then add version hint to it. Plus, the set returned from the utility may be immutable in the future. 

What about this?

```
List<String> otherMetadataFiles = Lists.newArrayList();
otherMetadataFiles.addAll(ReachableFileUtil.metadataFileLocations(table, false));
otherMetadataFiles.add(ReachableFileUtil.versionHintLocation(table));
return spark.createDataset(otherMetadataFiles, Encoders.STRING()).toDF("file_path");
```

**Review by aokolnychyi — APPROVED** [review]
Had a few minor nits. Looks great otherwise! Thanks for the work, @karuppayya!

File: core/src/main/java/org/apache/iceberg/ReachableFileUtil.java (hunk 14 above)
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/ReachableFileUtil.java]: Sounds good to me. Resolving this.

File: core/src/main/java/org/apache/iceberg/MetadataLocationUtils.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/MetadataLocationUtils.java]: Resolving this.

File: core/src/main/java/org/apache/iceberg/ReachableFileUtils.java (hunk 8 above)
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/ReachableFileUtils.java]: Resolving this.

File: core/src/main/java/org/apache/iceberg/MetadataLocationUtils.java [hunk 24]
```diff
@@ -0,0 +1,91 @@
...
+   * Returns all Metadata file paths which may not be in the current metadata. Specifically
+   * this includes "version-hint" files as well as entries in metadata.previousFiles.
+   * @param ops TableOperations of the table to get paths from
+   * @param isRecursive when true, recursively retrieves all the metadata.json files using metadata.previousFiles
+   * <p>when false, gets the all the metadata.json files pointed to by the only current metadata.json
+   * @return a list of paths to metadata files
+   */
+  public static List<String> miscMetadataFiles(TableOperations ops, boolean isRecursive) {
+    Set<String> miscMetadataLocation = new HashSet<>();
+    miscMetadataLocation.add(ops.metadataFileLocation("version-hint.text"));
```
[the remark this answers is not in the document]
Comment (aokolnychyi) [reply core/src/main/java/org/apache/iceberg/MetadataLocationUtils.java]: This has been addressed.

File: core/src/main/java/org/apache/iceberg/ReachableFileUtil.java (hunk 22 above)
Comment (karuppayya) [reply core/src/main/java/org/apache/iceberg/ReachableFileUtil.java:69]: org.apache.iceberg.TableMetadata.MetadataLogEntry is a static inner class of TableMetadata, I think Java does allow static import of class(Only methods and field members).

**Comment by aokolnychyi** [comment]
Thanks, @karuppayya! Thanks for reviewing, @flyrain @kbendick @RussellSpitzer!
